impl/spanner/testutil/testutil.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package testutil helps with tests against Spanner.
+package testutil
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"cloud.google.com/go/spanner/spannertest"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc"
+
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	databasepb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+var dbCount uint32
+
+func uniqueDBName() string {
+	database := fmt.Sprintf("fake-db-%d", atomic.AddUint32(&dbCount, 1))
+	const instance = "fake-instance"
+	const project = "fake-proj"
+	return fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, database)
+}
+
+// CreateDatabse returns a connection a 1 time use database with the ddl schema.
+func CreateDatabase(ctx context.Context, t testing.TB, ddlStatements []string) (*spanner.Client, func()) {
+	dbName := uniqueDBName()
+	client, adminClient, cleanup := inMemClient(ctx, t, dbName)
+	updateDDL(ctx, t, dbName, adminClient, ddlStatements...)
+	return client, cleanup
+}
+
+func inMemClient(ctx context.Context, t testing.TB, dbName string) (*spanner.Client, *database.DatabaseAdminClient, func()) {
+	t.Helper()
+	// Don't use SPANNER_EMULATOR_HOST because we need the raw connection for
+	// the database admin client anyway.
+
+	t.Logf("Using in-memory fake Spanner DB: %s", dbName)
+	srv, err := spannertest.NewServer("localhost:0")
+	if err != nil {
+		t.Fatalf("Starting in-memory fake: %v", err)
+	}
+	srv.SetLogger(t.Logf)
+	dialCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(dialCtx, srv.Addr, grpc.WithInsecure())
+	if err != nil {
+		srv.Close()
+		t.Fatalf("Dialing in-memory fake: %v", err)
+	}
+	client, err := spanner.NewClient(ctx, dbName, option.WithGRPCConn(conn))
+	if err != nil {
+		srv.Close()
+		t.Fatalf("Connecting to in-memory fake: %v", err)
+	}
+	adminClient, err := database.NewDatabaseAdminClient(ctx, option.WithGRPCConn(conn))
+	if err != nil {
+		srv.Close()
```

Comment:
Do we need to `conn.Close()` and `client.Close()` here? I wonder if there is an elegant way of doing all these `Close` calls (1, 2, 3, or 4 of them depending on where it fails) without duplicating them.

Reply:
I wish.  If there were a deferred object we could return that would help.